### PR TITLE
Add a Gather-Sum-Apply iteration method

### PR DESCRIPTION
**Pull request by balidani**
Hi!

This is my initial implementation for GSA. It seems to work with the greedy graph coloring example, but I will implement other examples as well to see whether it's flexible enough.

I have two questions:
- Right now I implemented the `gsa` method in `Graph`. Is there a more appropriate place? Should it be a `GraphAlgorithm` instead?
- Should we support different vertex types in the resulting graph?

Cheers!

### CODE REVIEW

**Comment by vasia** [comment]
Hi @balidani!
This is a great first step!

In my view, we should make this similar to Spargel and provide a method `runGatherSumApplyIteration` or `runGSAIteration` in `Graph`.
I would define 3 abstract classes: `GatherFunction` with a `gather` abstract method, `SumFunction` with a `sum` method and `ApplyFunction` with an `apply` method. These will be similar to Spargel's `MessagingFunction` and `VertexUpdateFunction` and will be wrapped by a map, a reduce and a join, respectively. 
The iteration can be wrapped inside a `GatherSumApplyIteration` class, similar to Spargel's `VertexCentricIteration`. This way we can later also add support for aggregators, broadcast sets and open/close methods.
We can move these classes inside their own package, `flink.graphs.gsa` or similar.
What do you think?

File: src/main/java/flink/graphs/Graph.java [hunk 1]
```diff
@@ -1321,4 +1312,75 @@ public void remove() {
...
+			MapFunction<Tuple3<Vertex<K, VV>, Edge<K, EV>, Vertex<K, VV>>, Tuple2<K, MO>> gather,
+			ReduceFunction<Tuple2<K, MO>> sum,
+			FlatJoinFunction<Tuple2<K, MO>, Vertex<K, VV>, Vertex<K, VV>> apply,
+			int maxIterations) {
+
+		DeltaIteration<Vertex<K, VV>, Vertex<K, VV>> iteration =
+				this.getVertices().iterateDelta(this.getVertices(), maxIterations, 0);
+
+		DataSet<Tuple3<Vertex<K, VV>, Edge<K, EV>, Vertex<K, VV>>> triplets = this
+				.getVertices()
```
Comment (vasia) [inline src/main/java/flink/graphs/Graph.java]: Retrieving the source vertices from the initial graph won't give you the updated vertex values, which will result into wrong behavior if your map-gather function uses the source vertex value. You can get the updated vertex values with `iteration.getSolutionSet()`.

**Comment by balidani** [comment]
Thanks for the feedback @vasia!

I'll start working on the changes.

**Comment by balidani** [comment]
I get a very strange exception when I change `getVertices` to `iteration.getSolutionSet`. Something about edges not being comparable. Here is the trace: https://gist.github.com/balidani/a95465ca53437e050ef2

It would be great to learn how to debug stuff in Flink, but right now I'm stumped. 

Any ideas? Thanks

**Comment by vasia** [comment]
Hi @balidani, 

I didn't thoroughly look into this, but I think that the problem is using "f0.f0" to do the key selection for the edge key.
It works if your project the edge key before doing the join, like this:

```
DataSet<Tuple3<Vertex<K, VV>, Edge<K, EV>, Vertex<K, VV>>> triplets = 
				iteration.getSolutionSet()
				.join(this.getEdges()
						.join(iteration.getWorkset())
						.where(1)
						.equalTo(0).with(new FlatJoinFunction<Edge<K, EV>, Vertex<K, VV>, 
								Tuple2<K, Tuple2<Edge<K, EV>, Vertex<K, VV>>>>() {

							public void join(Edge<K, EV> first,	Vertex<K, VV> second,
									Collector<Tuple2<K, Tuple2<Edge<K, EV>, Vertex<K, VV>>>> out) {

								out.collect(new Tuple2<K, Tuple2<Edge<K, EV>, Vertex<K, VV>>>(
										first.f0, new Tuple2<Edge<K, EV>, Vertex<K, VV>>(first, second)));
								
							}
						}))
				.where(0)
				.equalTo(0)
				.with(new TripletJoinFunction<K, VV, EV>());
```
and make sure to adjust `TripletJoinFunction` accordingly of course.

**Comment by balidani** [comment]
Thank you, this fixes the issue :)

**Comment by balidani** [comment]
Hi @vasia!

How exactly should I create a wrapper for the iterator?
Just by extending `DeltaIteration`? Or would we need a method on the vertex DataSet as well, something like `iterateGatherSumApply`?

Cheers!

**Comment by vasia** [comment]
@balidani you can follow `VertexCentricIteration.class` inside the spargel package, as a template for `GatherSumApplyIteration`.

**Pull request closed without merging**